### ISSUE TITLE
Kafka Receiver: Data Loss

### ISSUE THREAD

**Issue by pastequo** [cloudevents/sdk-go#643]
Hello again,

Continuing the (small) discussion here: https://github.com/cloudevents/sdk-go/issues/642

The client [here](https://github.com/cloudevents/sdk-go/blob/ab0ab9f7e0824a8ce40f394b8871de0e8465d725/v2/client/client.go#L203-L240) starts a new goroutine each time an event is pulled. At the end of "invoke" [here](https://github.com/cloudevents/sdk-go/blob/ab0ab9f7e0824a8ce40f394b8871de0e8465d725/v2/client/invoker.go#L42), Finish() is called. For kafka receiver it calls MarkMessage() [here](https://github.com/cloudevents/sdk-go/blob/ab0ab9f7e0824a8ce40f394b8871de0e8465d725/protocol/kafka_sarama/v2/receiver.go#L57)
This method is extracting the message offset, adding 1 to it and finally (not synchroneously) inform kafka that this consumer-group has processed all messages before message.offset+1

Since there is no control over the processing time (the time spent in invoke), the goroutine processing the last kafka message (for example) could finish first. Then the offset of this message is pushed "first", saying that this consumer group is up to date with the current kafka log for this partition. If the process is stopped / crashed / etc, before the other goroutine finished, when the process will be restarted (or when the rebalance is done for other consumer in the same group). Only new messages will be sent. All the other messages that were computed by the different goroutines are lost

For example, I pushed 100 messages in kafka in a single partition topic
After that, I start a consumer with this configuration
```golang
	saramaConfig := sarama.NewConfig()
	saramaConfig.Version = sarama.V2_0_0_0
	saramaConfig.Consumer.Offsets.AutoCommit.Enable = true
	saramaConfig.Consumer.Offsets.Initial = sarama.OffsetOldest
```

for each message, I apply a pseudo processing with a random duration
```
func receive(ctx context.Context, event cloudevents.Event) {
	// Random processing time
	n := rand.Intn(1000)
	time.Sleep(time.Duration(n) * time.Second)

	fmt.Printf("After %vs, %s\n", n, event.DataEncoded)
}
```
After a few second I stop the process using Ctrl-C
![image](https://user-images.githubusercontent.com/40832495/102256606-e12dd700-3f0b-11eb-8811-90aefaf98938.png)

Looking at the kafka offset I see
![image](https://user-images.githubusercontent.com/40832495/102256720-fefb3c00-3f0b-11eb-9641-80c9ffeafe6a.png)

Which is unexpected in many ways. I consumed 3 messages (id 26, 10 & 42). The current offset of my consumer group shouldn't be 100, should be 42+1

After investigating, there is an issue [here](https://github.com/cloudevents/sdk-go/blob/ab0ab9f7e0824a8ce40f394b8871de0e8465d725/protocol/kafka_sarama/v2/receiver.go#L57)
I think it should be something like 
```go
func (r *Receiver) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
	for message := range claim.Messages() {
		tmp := message
		m := NewMessageFromConsumerMessage(tmp)

		r.incoming <- msgErr{
			msg: binding.WithFinish(m, func(err error) {
				if protocol.IsACK(err) {
					session.MarkMessage(tmp, "")
				}
			}),
		}
	}
	return nil
}
```

(I would be happy to make a PR for this)

After applying this and replaying my scenario I have
![image](https://user-images.githubusercontent.com/40832495/102257331-e2abcf00-3f0c-11eb-9e56-b37ba606cb34.png)
and in kafka
![image](https://user-images.githubusercontent.com/40832495/102257453-0cfd8c80-3f0d-11eb-8036-b7bf8d19757d.png)

Which is better but still problematic im my opinion. I have only consume messages 9 & 25 and my offset is now 26. When my consumer will restart, the messages [0-8] U [10-24] won't be processed

I saw the notion of partitioning in cloudevents, maybe it would interesting to introduce it in the client (that should stay an abstraction above the different protocol) and consume partition by partition

I don't plan to use the Kafka Receiver, I don't have an use-case (I'm more likelyto follow @slinkydeveloper advices and only focus on transforming a sarama Message into a cloudevents and vice-versa). I just though it was important enough to mention it here

**Comment by slinkydeveloper** [comment]
Feel free to go ahead and PR the fix for the loop.

> Which is better but still problematic im my opinion. I have only consume messages 9 & 25 and my offset is now 26. When my consumer will restart, the messages [0-8] U [10-24] won't be consumed

This is weird, `MarkMessage` should not commit the offset immediatly but it should perform an offset tracking, that is it will commit 7 only when all messages from 0 to 6 are marked... Maybe I'm missing something here...

**Comment by pastequo** [comment]
I think this is not the right behavior, it doesn't have this logic

https://github.com/Shopify/sarama/blob/0616f68815691527d3b6ab4d95a2881e09400fe3/offset_manager.go#L515-L519

**Comment by slinkydeveloper** [comment]
Ah... well, then it doesn't work for sure :smile: We probably might need to implement by ourselves an offset tracker...

**Comment by pastequo** [comment]
... Or ensure that message are consumed & processed 1 by 1 per partition
I understand that the Client has to be an abstraction above protocols, but would it make sense to introduce the notion of partition into it ?

This notion is already present in the spec
https://github.com/cloudevents/spec/blob/v1.0/extensions/partitioning.md

**Comment by slinkydeveloper** [comment]
> I understand that the Client has to be an abstraction above protocols, but would it make sense to introduce the notion of partition into it ?

I don't think so, to me this is outside the scope of the Client. It's not simple to abstract the partition concept and it will be useful only for kafka anyway.
